File: sample-projects/kotlin/app/src/main/kotlin/com/example/app/UserService.kt
package com.example.app

import com.example.core.Entity
import com.example.core.Identifiable
import com.example.core.Result
import com.example.core.capitalizeWords
import com.example.core.getOrNull
import com.example.models.Permission
import com.example.models.User
import com.example.models.UserRepository

class UserService(private val repository: UserRepository) {

    fun createUser(name: String, email: String, permission: Permission): Result<User> {
        val normalizedName = name.capitalizeWords()
        val user = User(
            id = System.currentTimeMillis(),
            name = normalizedName,
            email = email,
            permission = permission
        )
        return repository.save(user)
    }

    fun getUser(id: Long): User? = repository.findById(id).getOrNull()

    fun listAdmins(): List<User> =
        repository.findAll().filter { it.permission == Permission.ADMIN }

    fun <T> validateEntity(entity: Entity): Boolean = entity.isValid()

    fun <T> describeIdentifiable(item: Identifiable<T>): String =
        "Item with id=${item.id}"
}

File: sample-projects/ruby/lib/app/user_service.rb
# typed: strict
# frozen_string_literal: true

require_relative '../core/string_utils'
require_relative '../core/result'
require_relative '../models/user'
require_relative '../models/user_repository'
require_relative '../models/permission'

module App
  class UserService
    extend T::Sig

    sig { params(repository: Models::UserRepository).void }
    def initialize(repository)
      @repository = T.let(repository, Models::UserRepository)
      @next_id = T.let(1, Integer)
    end

    sig { params(name: String, email: String, permission: Models::Permission).returns(Core::Result) }
    def create_user(name, email, permission)
      formatted_name = Core::StringUtils.capitalize_words(name)
      user = Models::User.new(@next_id, formatted_name, email, permission)
      result = user.validate
      if result.success?
        @repository.save(user)
        @next_id += 1
      end
      result
    end

    sig { returns(T::Array[Models::User]) }
    def list_admins
      @repository.find_all.select { |user| user.permission == Models::Permission::Admin }
    end
  end
end
